### PR TITLE
NS-46 Partial replacement of SIS Enrollments API by Data Loch

### PR DESCRIPTION
**Pull request by raydavis**
https://jira.ets.berkeley.edu/jira/browse/NS-46

Apologies for how massive this is. The SIS Enrollments API weaved itself through a lot of code.

The only UX change should be that non-primary section drops are no longer included in the dropped classes list. Secondary section drops are always either redundant (because of being tied to a primary drop), or just swaps to get into another secondary section. They don't reflect thrown-away effort, as primary drops usually do. I'll ping Flint on the JIRA ticket, too.

See BOAC-905 for the followup.

### CODE REVIEW

File: boac/merged/sis_sections.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+"""
+
+
+import re
+
+from boac.externals import data_loch
+from boac.models.json_cache import stow
+
+
+@stow('sis_section_{sis_section_id}', for_term=True)
```
Comment (raydavis) [inline boac/merged/sis_sections.py]: This replaces one chunk of NormalizedCacheEnrollment code-and-storage.

File: boac/externals/data_loch.py [hunk 2]
```diff
@@ -92,6 +92,44 @@ def _get_course_enrollments(course_id):
...
+                  crs.sis_course_title, crs.sis_course_name,
+                  crs.sis_section_id, crs.sis_primary, crs.sis_instruction_format, crs.sis_section_num
+              FROM {intermediate_schema()}.sis_enrollments enr
+              JOIN {intermediate_schema()}.course_sections crs
+                  ON crs.sis_section_id = enr.sis_section_id
+                  AND crs.sis_term_id = enr.sis_term_id
+              WHERE enr.ldap_uid = {uid}
+                  AND enr.sis_enrollment_status != 'D'
+                  AND enr.sis_term_id = {term_id}
+              ORDER BY crs.sis_course_name, crs.sis_primary DESC, crs.sis_instruction_format, crs.sis_section_num;
```
Comment (johncrossman) [inline boac/externals/data_loch.py]: @raydavis  minor: unnecessary trailing `;`

File: boac/merged/sis_sections.py (hunk 1 above)
Comment (johncrossman) [reply boac/merged/sis_sections.py]: Chipping away! Nice.

**Comment by johncrossman** [comment]
Looks solid, @raydavis  And thx for test coverage. I'll hand it over to the eyes of @sandeepmjay

File: boac/externals/sis_enrollments_api.py [hunk 3]
```diff
@@ -28,11 +28,36 @@
...
+    dropped_classes = []
+    midterm_grades = {}
+    for enrollment in enrollments:
+        enrollment_status = enrollment.get('enrollmentStatus', {}).get('status', {}).get('code')
+        if enrollment_status == 'D' and enrollment.get('gradingBasis', {}).get('code') != 'NON':
+            dropped_classes.append({
+                'displayName': enrollment.get('classSection', {}).get('class', {}).get('course', {}).get('displayName'),
+                'component': enrollment.get('classSection', {}).get('component', {}).get('code'),
+                'sectionNumber': enrollment.get('classSection', {}).get('number'),
+                'dropDate': enrollment.get('enrollmentStatus', {}).get('fromDate'),
```
Comment (paulkerschen) [inline boac/externals/sis_enrollments_api.py]: @raydavis Unfortunately, we have some doubts that this `fromDate` actually indicates a date dropped, as we’d like it too. See https://jira.ets.berkeley.edu/jira/browse/BOAC-392...

Comment (raydavis) [reply boac/externals/sis_enrollments_api.py]: Thank you @paulkerschen , and I apologize for the typing. I went looking for some evidence of an issue but didn't manage to hit the proper set of search items.

**Comment by sandeepmjay** [comment]
Looks good to me @raydavis apart from the data quality question @paulkerschen had for dropDate.

**Comment by raydavis** [comment]
Thanks, all. PR amended.

**Comment by sandeepmjay** [comment]
Waiting for Mr.T to green light it.

**Pull request merged**